The error occurs because the `table_exists` method in the `HiveCommandClient` class does not correctly handle cases where the `partition` parameter is not None. The method is trying to call the `partition_spec` method incorrectly, resulting in incorrect behavior.

The issue resides in the line:
`stdout = run_hive_cmd("""use %s; show partitions %s partition (%s)""" % (database, table, self.partition_spec(partition)))`

The bug occurs because the `partition_spec` is expected to be called as an instance method of the object, but it is used as if it were a class method.

To fix the bug, you should remove the `self` parameter from the `partition_spec` function definition and calls since it's an instance method. 

The corrected code for the `table_exists` method is as follows:

```python
def table_exists(self, table, database='default', partition=None):
    if partition is None:
        stdout = run_hive_cmd('use {0}; show tables like "{1}";'.format(database, table))

        return stdout and table in stdout
    else:
        stdout = run_hive_cmd("""use %s; show partitions %s partition (%s)""" % (database, table, partition_spec(partition)))

        if stdout:
            return True
        else:
            return False
```

In this corrected implementation, we remove the incorrect use of `self` when calling the `partition_spec` function. This change should resolve the bug by properly calling the `partition_spec` method as an instance method.